0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+
+#### Give `Store` first-class support as an alternative to `CircuitInstruction`
+
+The RFC currently proposes to jam `Store` into being a subclass of `Instruction` that has inherent classical bits (similar to `Instruction.condition` or the `ControlFlowOp` subclasses).
+This means that all consumers of `QuantumCircuit` that care about data-flow orderings and display must know to check for this particular subclass, and include custom code to handle this.
+Similarly, the `DAGCircuit` would need to grow an alternative to `DAGOpNode` that could represent this with new first-class support.
+
+A strong alternative would be to have different versions of the `CircuitInstruction`/`DAGOpNode` circuit context object that better represent the "special" classical operations such as `Measure`, `Store` and the `ControlFlowOp` subclasses.
+This does not inherently change that the user might forget to handle these cases, but changing the structure of the `CircuitInstruction` objects as they appear in Python space offers us a way to make them fail loudly, not silently ignore data dependencies.
+In Rust space, we can use the algebraic type system to more easily enforce that all cases are handled.
+I think this is a direction we absolutely must take in the future, but that the time to do that is once the backing data of both `QuantumCircuit` and `DAGCircuit` is stored in Rust space and the transition path is easier.
+If we were to do this right now, all consumers of `QuantumCircuit.data` would be broken, and this would either require an immediate huge amount of work to change how `Measure` and the control-operations are stored in Qiskit, or we would be adding an even more special case that `Store` already is in that it would not just be a special `Instruction`, but also a special `CircuitInstruction`.
+I propose that we revisit the internal representation of all these classical operations in one unified redesign for both `QuantumCircuit` and `DAGCircuit`, rather than trying to do it on-the-fly.
+
+
+### Alternatives to smaller details
+
+#### Automatic closure over variables declared as `input`
+
+In the current system, it is not valid to have a circuit that has both closure variables _and_ input or output variables.
+Instead of having separate `add_input()` and `add_capture()` methods, one could imagine that inner scopes could `add_var` variables to be either of type "input" or a new type "closure".
+The reason to choose to have a separate method for these semantics is that the signature of the `add_var` method would become quite unpleasant if it needed to do all these things; it already has multiple signatures because it needs to take either `(var,)` or `(name, type)`.
+
+It is not possible to close over a variable with just its name and type, because we'd also need the information about which root allocation it refers to to build up the `DAGCircuit` data-flow graph, and we don't gain that information until the circuit is embedded within another.
```

Comment:
```suggestion
It is not possible to close over a variable with just its name and type, because we'd also need the information about which root allocation it refers to build up the `DAGCircuit` data-flow graph, and we don't gain that information until the circuit is embedded within another.
```